### ISSUE TITLE
[AArch64] Reductions of `<N x i1>` vectors can lead to miscompilations

### ISSUE THREAD

**Issue by Sp00ph** [llvm/llvm-project#62211]
This code leads to the miscompilation:
```ll
declare i1 @llvm.vector.reduce.or(<2 x i1>)

define i1 @bad(<2 x i32> %a) noinline {
  %b = trunc <2 x i32> %a to <2 x i1>
  %ret = call i1 @llvm.vector.reduce.or(<2 x i1> %b)
  ret i1 %ret
}

define i32 @main() {
    %a = call i1 @bad(<2 x i32> <i32 1, i32 2>)
    %b = zext i1 %a to i32
    ret i32 %b
}
```
Compiling this with `-O3` leads to the following assembly:
```asm
bad:                                    // @bad
        umaxp   v0.2s, v1.2s, v1.2s
        fmov    w8, s0
        and     w0, w8, #0x1
        ret
.LCPI1_0:
        .word   1                               // 0x1
        .word   2                               // 0x2
main:                                   // @main
        str     x30, [sp, #-16]!                // 8-byte Folded Spill
        adrp    x8, .LCPI1_0
        ldr     d1, [x8, :lo12:.LCPI1_0]
        bl      bad
        and     w0, w0, #0x1
        ldr     x30, [sp], #16                  // 8-byte Folded Reload
        ret
```
which when executed returns a 0 from `main`. After removing the `noinline`, a 1 is returned from `main` instead.

**Comment by Sp00ph** [comment]
Opened https://reviews.llvm.org/D148672 which should fix this issue.

**Comment by topperc** [comment]
Fixed by 3c9083f6757cbaf6f8d6c601586d99a11faf642e
